feat: add unified API documentation (#368)
amp-labs/docs: src/unified-api.mdx

<CodeGroup>
```yaml Jira
- objectName: issue
  mapToName: ticket
  requiredFields:
    - fieldName: summary
      mapToName: title
    - fieldName: status
      mapToName: status
```

```yaml Linear
- objectName: issues
  mapToName: ticket
  requiredFields:
    - fieldName: title
      mapToName: title
    - fieldName: state
      mapToName: status
```

```yaml Zendesk
- objectName: tickets
  mapToName: ticket
  requiredFields:
    - fieldName: subject
      mapToName: title
    - fieldName: status
      mapToName: status
```

```yaml Result
{
  "object": "ticket",
  "data": {
    "title": "Ticket 1",
    "status": "open"
  }
}
```
</CodeGroup>

## Migration from merge.dev

<Steps>
  <Step title="Use Compatible Templates">
    Start with our merge.dev-compatible templates:
    ```bash
    # CRM template (matches merge.dev's model)
    amp deploy samples/unifiedCRM
    
    # Ticketing template
    amp deploy samples/unifiedTicketing
    ```
  </Step>
  
  <Step title="Update Your Code">
    Replace merge.dev SDK calls:
    ```javascript
    // merge.dev
    const contacts = await merge.crm.contacts.list()
    
    // Ampersand

>     const contacts = await ampersand.read('contact')

Author: Just updated the code to match Ampersand's approach

---
title: "Build your own unified API"
sidebarTitle: "Unified API"
---

## What is a unified API?

A unified API provides a single, consistent interface to interact with multiple third-party services. Instead of learning each provider's unique schema, you define one schema that works across all of them.

> <Note>

Reviewer: this doesn't need to be in a "Note", we save those for special call outs.

Author: done on https://github.com/amp-labs/docs/pull/368/commits/18dd1324ec7ecade609077730e5438ba05d5c52a

This guide covers one of Ampersand's use cases: building unified APIs. You can also use Ampersand for provider-specific integrations or mix both approaches. Unlike traditional unified API providers that force you into their predefined models, Ampersand lets you design your own schema.
</Note>

> ## Why build a unified API?

Reviewer: @caiopizzol can you remove this section for "Why build a unified API?"  It's a bit too "market-y" for docs.

> - **Faster development** - Add providers without changing application logic

Reviewer: ```suggestion
- **Faster development** - Add new integrations without changing application logic
```

- **Consistent data model** - One schema regardless of source

> - **Real-time updates** - Subscribe to field-level changes

Author: done on https://github.com/amp-labs/docs/pull/368/commits/386a1ddbcb96a4df82b0c13e58e042f0e68740a4

- **No vendor lock-in** - You control the schema and where data lives

## How it works

### 1. Map objects and fields

```yaml
# Map different provider objects to unified names
- objectName: account        # Salesforce
  mapToName: company
  requiredFields:
    - fieldName: name
      mapToName: company_name
    - fieldName: annualrevenue
      mapToName: revenue

- objectName: companies      # HubSpot
  mapToName: company
  requiredFields:
    - fieldName: name
      mapToName: company_name
    - fieldName: annualrevenue
      mapToName: revenue
```

### 2. Receive unified + raw data

Every webhook includes both your unified fields and the complete provider response:

```json
{
  "mappedFields": {
    "remote_id": "001abc",
    "company_name": "Acme Corp",
    "revenue": 5000000,
    "remote_created_at": "2024-01-15",
    "remote_updated_at": "2024-08-23"
  },
  "raw": {
    "Id": "001abc",
    "Name": "Acme Corp",
    "AnnualRevenue": 5000000,
    "CreatedDate": "2024-01-15T10:30:00Z",
    "SystemModstamp": "2024-08-23T14:20:07Z",
    "Custom_Field__c": "value"  // All fields available
  }
}
```

> <Tip>

Reviewer: you can remove this tip, since it's a minor point and already covered in the table below

Unlike merge.dev where you need `include_remote_data=true`, Ampersand always includes raw data. This means you can use your unified schema while still accessing provider-specific fields.
</Tip>

### 3. Real-time updates with subscribe actions

Get webhooks when data changes (1-2 minutes latency):

```yaml
subscribe:
  objects:
    - objectName: account
      destination: webhook
      inheritFieldsAndMapping: true
      updateEvent:
        enabled: always
        requiredWatchFields:
          - company_name  # Your unified field
          - revenue
```

```json
// Webhook payload on update
{
  "subscribeEventType": "update",
  "mappedFields": {
    "company_name": "Acme Corp Updated",
    "revenue": 6000000
  },
  "raw": { /* full provider data */ }
}
```

## Key architectural differences from merge.dev

### Data persistence

**Ampersand:** No data storage. Data streams directly from providers to your webhooks. You control where and how to persist data.

**Merge.dev:** Stores all customer data in their cloud by default. Enables faster syncs but requires them as a data sub-processor. "Merge Destinations" (Enterprise) bypasses storage.

> **Why it matters:** For healthcare, finance, or regulated industries, Ampersand's passthrough architecture eliminates compliance concerns around third-party data storage.

Reviewer: ```suggestion
**Why it matters:** For healthcare, finance, or regulated industries, Ampersand's transport-only architecture eliminates compliance concerns around third-party data storage.
```

### Two-way sync

**Ampersand:**
- **Read**: Scheduled syncs OR real-time subscribe actions (field-level granularity)
- **Write**: Direct write actions with field mapping
- **Pattern**: Subscribe + Write = true bidirectional real-time

**Merge.dev:**
- **Read**: Scheduled syncs + poll after webhook notifications (model-level)

> - **Write**: Instant POST/PATCH operations

Reviewer: is this true? does merge directly make the write to the provider right away? also do you know if they apply mappings in their writes? if not, then that's worth highlighting. 

I'd also remove the word "instant", it sounds too positive.

> - **Pattern**: Polling + Writes (third-party webhooks on Pro/Enterprise)

Reviewer: ```suggestion
- **Pattern**: Polling + writes
```

### Comparison table

| Feature | Ampersand | Merge.dev |
|---------|-----------|-----------|
| **Schema** | You define it | Fixed common model |

> | **Raw data** | Always included | `include_remote_data=true` flag |

Reviewer: ```suggestion
| **Raw data** | Always included | need `include_remote_data=true` flag |
```

> | **Data storage** | None (you control) | Stored in Merge cloud |

Reviewer: ```suggestion
| **Data storage** | No persistent storage | Stored in Merge cloud |
```

| **Real-time** | Subscribe actions (all plans) | Third-party webhooks (Pro/Enterprise) |

> | **Event detail** | Field-level changes | Model-level notifications |

Reviewer: ```suggestion
| **Event granularity** | Field-level changes | Model-level notifications |
```

> | **Sub-processor** | No | Yes |

Reviewer: can you remove this? we are technical still a data sub-processor since the data passes through us

> ## Examples

Author: remove on https://github.com/amp-labs/docs/pull/368/commits/ac9b6e6c4ad584cb9c473258c1cd21e0a90f5230

    "assignee": { "id": "user-456" }
  }
}
```
</CodeGroup>

## Migration from merge.dev

<Steps>
  <Step title="Understand the persistence model">
    **Merge.dev:** Stores your customer data; you poll their API  
    **Ampersand:** Streams data to your webhooks; you store it
    
    Set up webhook endpoints and decide on your storage strategy (database, warehouse, etc.)
  </Step>

  <Step title="Use compatible templates">
    ```bash
    git clone https://github.com/amp-labs/samples.git
    cd samples/unifiedCRM  # or unifiedTicketing
    amp deploy --project=my-project
    ```
  </Step>
  
  <Step title="Update webhook handlers">
    ```javascript
    // merge.dev: Poll after webhook notification
    app.post('/merge-webhook', async (req, res) => {
      const contacts = await merge.crm.contacts.list({
        modified_after: lastSync,
        include_remote_data: true
      });
      await db.contacts.bulkUpsert(contacts);
      res.sendStatus(200);
    });
    
    // Ampersand: Data arrives in webhook
    app.post('/webhook/contacts', async (req, res) => {
      const { mappedFields, raw } = req.body.result;
      await db.contacts.upsert({
        ...mappedFields,

>         providerData: raw

Reviewer: add `...fields,`

- **Consistent data model** - Your application works with one schema regardless of the source
- **Gradual migration** - Start with compatibility mode, customize as needed
- **No vendor lock-in** - You control the schema, not limited by lowest common denominator

## Core concepts

### Object mapping
Map different provider objects to a unified name:

```yaml
# Salesforce Account → company
# HubSpot companies → company  
# Pipedrive organizations → company

- objectName: account        # Salesforce
  mapToName: company
- objectName: companies      # HubSpot
  mapToName: company
- objectName: organizations  # Pipedrive
  mapToName: company
```

### Field standardization
Create consistent field names across providers:

```yaml
# Different providers, same unified fields
requiredFields:
  - fieldName: name           # Salesforce
    mapToName: company_name
  - fieldName: name           # HubSpot
    mapToName: company_name
  - fieldName: account_name   # Zoho
    mapToName: company_name
```

### Raw data access

Every payload includes both unified fields AND the original provider data:

```json
{

>   "mappedFields": {

Author: done in: https://github.com/amp-labs/docs/pull/368/commits/f7bda862e73d9a297cd0ebd3215ec7a1452d6d2d
